internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriter.java

```diff
@@ -1994,18 +1989,25 @@ private Segment startSegment(int partition) {
 
   private void endAllSegments(boolean finalSegment) {
     Segment segment;
+    List<CompletableFuture> futures = new ArrayList<>();
     for (int i = 0; i < segments.length; i++) {
       segment = segments[i];
       if (segment != null) {
-        endSegment(i, finalSegment);
+        futures.add(endSegment(i, finalSegment));
       }
     }
+    try {
+      CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()])).get();
+    } catch (ExecutionException | InterruptedException e) {
+      handleControlMessageProducingException(e, ControlMessageType.END_OF_SEGMENT);
+    }
   }
 
   /**
    * @param partition in which to end the current segment
+   * @return A CompletableFuture which either contains a result if a EOS was sent, or null if none needed to be sent.
    */
-  public void endSegment(int partition, boolean finalSegment) {
+  public CompletableFuture<PubSubProduceResult> endSegment(int partition, boolean finalSegment) {
```

Comment:
There are some places that expect this call to be a blocking call. Should be replace it them with `future.get()`?

Reply:
Sure, I will take a look at each one.